File: src/pipeline/detectors/shell_command_not_found.py
"""
Detector for shell errors when a command/script is not found.
"""

import re
import typing as T
from src.pipeline.detectors.base import Detector
from src.pipeline.models import ErrorClue


class ShellCommandNotFoundDetector(Detector):
    """
    Detect shell errors when a command/script is not found.

    Matches patterns like:
    - ./test.sh: line 3: ./configure: No such file or directory
    - ./test.sh: 2: ./configure: not found
    - /bin/sh: ./script.sh: not found
    """

    PATTERNS = {
        "missing_file": r":\s*line\s+\d+:\s*\.?/?(?P<file_path>[^\s:]+):\s*No such file or directory",
        "missing_file_not_found": r":\s*\d+:\s*\.?/?(?P<file_path>[^\s:]+):\s*not found",
    }

    EXAMPLES = [
        (
            "./test.sh: line 3: ./configure: No such file or directory",
            {
                "clue_type": "missing_file",
                "confidence": 1.0,
                "context": {"file_path": "configure"},
            },
        ),
        (
            "./test.sh: 2: ./script.sh: not found",
            {
                "clue_type": "missing_file_not_found",
                "confidence": 1.0,
                "context": {"file_path": "script.sh"},
            },
        ),
    ]

File: src/boil/detectors/shell_command_not_found.go
// Detector for shell command not found errors.
package detectors

// ShellCommandNotFoundDetector detects shell command not found errors
type ShellCommandNotFoundDetector struct {
	*BaseDetector
}

func NewShellCommandNotFoundDetector() (*ShellCommandNotFoundDetector, error) {
	base, err := NewBaseDetector(
		"ShellCommandNotFoundDetector",
		100,
		map[string]string{
			"missing_file":           `:\s*line\s+\d+:\s*\.?/?(?P<file_path>[^\s:]+):\s*No such file or directory`,
			"missing_file_not_found": `:\s*\d+:\s*\.?/?(?P<file_path>[^\s:]+):\s*not found`,
		},
		[]DetectorExample{
			{
				Name:     "shell_no_such_file",
				Input:    "bash: line 5: ./run.sh: No such file or directory",
				ClueType: "missing_file",
				Context:  map[string]string{"file_path": "run.sh"},
			},
			{
				Name:     "shell_not_found",
				Input:    "sh: 1: ./script: not found",
				ClueType: "missing_file_not_found",
				Context:  map[string]string{"file_path": "script"},
			},
		},
	)
	if err != nil {
		return nil, err
	}
	return &ShellCommandNotFoundDetector{base}, nil
}
